### PR TITLE
Fix recently used documents on Linux and Windows

### PR DESCRIPTION
**Pull request by fxha**
| Q                | A
| ---------------- | ---
| Bug fix?         | yes
| Fixed tickets    | #130
| License          | MIT

### CODE REVIEW

**Comment by Jocs** [comment]
@fxha Can you show me some pictures of the `recent files`. Because I have no windows computers.

File: src/main/menu.js [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    recentUsedDocuments = getRecentlyUsedDocuments()
+  }
+
+  // "we don't support changing menu object after calling setMenu, the behavior
+  // is undefined if user does that." That means we have to recreate the
+  // application menu each time.
+
+  const menu = Menu.buildFromTemplate(configureMenu(recentUsedDocuments))
+  Menu.setApplicationMenu(menu)
+  if (!this.init && process.platform === 'darwin') {
```
Comment (Jocs) [inline src/main/menu.js]: Sorry, I don't find where are you defined `init` ? or what's `this` refer to?

File: src/main/menu.js [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+export const getRecentlyUsedDocuments = () => {
+  if (!isFile(recentlyUsedDocumentsPath)) {
+    return []
+  }
+
+  try {
+    let recentDocuments = JSON.parse(fs.readFileSync(recentlyUsedDocumentsPath, 'utf-8'))
+    for (const item of recentDocuments) {
```
Comment (Jocs) [inline src/main/menu.js]: maybe `Array.prototype.filter` is a better method to to this.

File: src/main/menu.js [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+      if (!item || !isFile(item)) {
+        recentDocuments.splice(recentDocuments.indexOf(item), 1)
+      }
+    }
+    if (recentDocuments.length > MAX_RECENTLY_USED_DOCUMENTS) {
```
Comment (Jocs) [inline src/main/menu.js]: is it needed?

File: src/main/actions/file.js [hunk 4]
```diff
@@ -201,8 +202,14 @@ export const open = win => {
     }]
   })
   if (filename && filename[0]) {
-    const newWindow = createWindow(filename[0])
-    watchAndReload(filename[0], newWindow)
+    openDocument(win, filename[0])
```
Comment (Jocs) [inline src/main/actions/file.js]: Define `openDocument` function before use it.

File: src/main/actions/file.js [hunk 5]
```diff
@@ -236,3 +243,7 @@ export const moveTo = win => {
 export const rename = win => {
   win.webContents.send('AGANI::ask-file-rename')
 }
+
+export const clearRecentlyUsed = win => {
```
Comment (Jocs) [inline src/main/actions/file.js]: `win` is not used.

File: src/main/menus/file.js [hunk 6]
```diff
@@ -3,31 +3,62 @@ import * as actions from '../actions/file'
...
-    click (menuItem, browserWindow) {
-      actions.newFile()
+  if (notOsx) {
+    let recentlyUsedMenu = {
+      label: 'Open Recent',
+      submenu: [{
+        label: 'Clear Recently Used',
+        enabled: recentlyUsedFiles.length > 0,
+        click (menuItem, browserWindow) {
+          actions.clearRecentlyUsed(browserWindow)
```
Comment (Jocs) [inline src/main/menus/file.js]: `browserWindow` is not used.

File: src/main/menu.js [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+const recentlyUsedDocumentsPath = path.join(getPath('userData'), FILE_NAME)
+const isOsxOrWindows = /darwin|win32/.test(process.platform)
+
+export const addRecentlyUsedDocuments = filePath => {
+  if (isOsxOrWindows) app.addRecentDocument(filePath)
+  if (process.platform === 'darwin') return
+
+  let recentDocuments = getRecentlyUsedDocuments()
+  const index = recentDocuments.indexOf(filePath)
+  let needSave = index !== 0
```
Comment (Jocs) [inline src/main/menu.js]: I am a little confused.

>if (index !== 0)

is it meaning that filePath is not in `recentDocuments` or this index great that `0`.

I think if `index` equal `-1`, then we need to save it to  `recently-used-documents.json` ?

**Comment by fxha** [comment]
@Jocs Preview: https://streamable.com/m2f2v

And there is a separator between `Clear Recently Used` and recently used documents - electron seems to have problems with my theme/gdm or wayland.

File: src/main/menu.js (hunk 1 above)
Comment (fxha) [reply src/main/menu.js]: `this.init` is defined at the bottom of the function. It's first undefined and after first call `true`

File: src/main/menu.js (hunk 3 above)
Comment (fxha) [reply src/main/menu.js]: Yes of course, if you manipulate the file

File: src/main/menu.js (hunk 7 above)
Comment (fxha) [reply src/main/menu.js]: It means the file is not the last recently used file. The file does not exist in the list or it exists, but is not the last recently used file.

File: src/main/menus/file.js [hunk 8]
```diff
@@ -3,31 +3,62 @@ import * as actions from '../actions/file'
...
-  submenu: [{
-    label: 'New File',
-    accelerator: 'CmdOrCtrl+N',
-    click (menuItem, browserWindow) {
-      actions.newFile()
+  if (notOsx) {
+    let recentlyUsedMenu = {
+      label: 'Open Recent',
+      submenu: [{
+        label: 'Clear Recently Used',
```
Comment (Jocs) [inline src/main/menus/file.js]: Please put the `Clear Recently Used` at the bottom of menu.(it means the last menu item).

File: src/main/menu.js [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+  // is undefined if user does that." That means we have to recreate the
+  // application menu each time.
+
+  const menu = Menu.buildFromTemplate(configureMenu(recentUsedDocuments))
+  Menu.setApplicationMenu(menu)
+  if (!this.init && process.platform === 'darwin') {
+    // app.dock is only for macosx
+    app.dock.setMenu(dockMenu)
+  }
+  this.init = true
```
Comment (Jocs) [inline src/main/menu.js]: But it not a `Class` or `object`, it's a function, so what `this` refer to?

File: src/main/menu.js (hunk 7 above)
Comment (Jocs) [reply src/main/menu.js]: Ok, I need think again. and you are right!

File: src/main/menu.js (hunk 9 above)
Comment (Jocs) [reply src/main/menu.js]: Yes, `function` is `object` too.

But, I think declare a variable is more readable.

```javascript
let isInit = false
// do something
if (isInit) {}
```

Comment (fxha) [reply src/main/menu.js]: So then we have to declare it outside the function? Otherwise it's always `false`

Comment (Jocs) [reply src/main/menu.js]: Yeh, outside the function.

**Comment by fxha** [comment]
@Jocs  Updated PR

**Pull request merged**